tests/test_cache.py

```diff
@@ -0,0 +1,98 @@
+    cacher = FileCache()    
+    return cacher
+
+
+@pytest.fixture()
+def non_empty_cache(fs):
+    for path in CONFIG.values():
+        fs.create_dir(path)
+
+    cache = FileCache()    
+    fs.create_file(Path(cache.cache_dir, TEST_FNAME))
+    return cache
+
+def assert_file_in_index(cache, expected_data):
+    """Utility function to check that the file is in the index.
+    Only works if it's the first non header row"""
+    with open(cache.cache_index, "r") as fhandle:
+        data = csv.DictReader(fhandle)
+        assert next(data) == expected_data
+
+def test_init(fs):
+    """Tests that the cache and remote directories are set properly"""
+    cache_dir = "/some/test/dir"
+    fs.create_dir(cache_dir)
+    remote = "https://www.github.com"
+    cache = FileCache(cache_dir, remote)
+    assert cache.cache_dir == cache_dir
+    assert cache.remote_url == remote
+    assert Path(cache_dir, CACHE_INDEX).exists()
+
+
+def test_get_when_not_in_cache(fs, cache):
+    """Tests that when a file is not in the cache, that it is fetched"""
+    fname = "test.txt"
+    fs.create_file(Path(cache.remote_url, fname))
+    cache.get(fname)
+    expected = {
+        CACHE_HEADER[0]: str(Path(cache.cache_dir, fname)),
+        CACHE_HEADER[1]: str(Path(cache.remote_url, fname))
+    }
+    assert Path(cache.cache_dir, fname).exists()
+    assert_file_in_index(cache, expected)
+
+
+@mock.patch.object(FileCache, "handle_local_file")
+def test_get_when_cached(mock, non_empty_cache):
+    """Tests that when a file is cached, it is not fetched using _get_file()"""
+    cache_path = non_empty_cache.get(TEST_FNAME)
+    expected = {
+        CACHE_HEADER[0]: str(Path(non_empty_cache.cache_dir, TEST_FNAME)),
+        CACHE_HEADER[1]: str(Path(non_empty_cache.remote_url, TEST_FNAME))
+    }
+    assert mock.call_count == 0
+    assert cache_path == str(Path(non_empty_cache.cache_dir, TEST_FNAME))
+    assert_file_in_index(non_empty_cache, expected)
+
+
+def test_in_cache_index_but_not_dir(fs, cache):
+    """Tests that when a file is in the cache index, but not actually in the
+    cache directory, that it is fetched"""
+    fname = "fake_file.txt"
+    # sabotage the index by adding to it
+    with open(cache.cache_index, 'a') as fhandle:
+        writer = csv.DictWriter(fhandle, fieldnames=CACHE_HEADER)
+        writer.writerow({
+            CACHE_HEADER[0]: str(Path(cache.cache_dir, fname)),
+            CACHE_HEADER[1]: "/random/directory"
+        })
+    with mock.patch.object(FileCache, "handle_local_file") as mocked_function:
+        cache.get("fake_file.txt")
+        assert mocked_function.call_count == 1
+
+
+# not sure why this is failing now
+# OSError: Could not find a suitable TLS CA certificate bundle, invalid path: /Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/certifi/cacert.pem
+def test_remote_download(cache):
+    remote_file = "https://github.com/tammy-dg/tammys-simple-web-server/blob/master/test/test.png"
```

Comment:
weird - this is a public repo. debug together?

Reply:
We think it's caused by pytest.
Full traceback:
```
_____________________________________________________________________________________ test_remote_download _____________________________________________________________________________________

cache = <dnadoh.cache.FileCache object at 0x7fdeadaf5580>

    @pytest.skip("Failing test due to certificate issues")
    def test_remote_download(cache):
>       remote_file = "http://github.com/tammy-dg/tammys-simple-web-server/blob/master/test/test.png?raw=true"

/Users/tammylau/workspace/personal_devel/DNA-doh/tests/test_cache.py:94: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/tammylau/workspace/personal_devel/DNA-doh/dnadoh/cache.py:89: in get
    self._download_file(requested_file, cache_path)
/Users/tammylau/workspace/personal_devel/DNA-doh/dnadoh/cache.py:111: in _download_file
    h = requests.head(remote_url, allow_redirects=True)
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/api.py:102: in head
    return request('head', url, **kwargs)
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/api.py:61: in request
    return session.request(method=method, url=url, **kwargs)
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/sessions.py:529: in request
    resp = self.send(prep, **send_kwargs)
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/sessions.py:667: in send
    history = [resp for resp in gen]
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/sessions.py:667: in <listcomp>
    history = [resp for resp in gen]
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/sessions.py:237: in resolve_redirects
    resp = self.send(
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/sessions.py:645: in send
    r = adapter.send(request, **kwargs)
/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/adapters.py:417: in send
    self.cert_verify(conn, request.url, verify, cert)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <requests.adapters.HTTPAdapter object at 0x7fdeada91bb0>, conn = <urllib3.connectionpool.HTTPSConnectionPool object at 0x7fdeadaccb50>
url = 'https://github.com/tammy-dg/tammys-simple-web-server/blob/master/test/test.png?raw=true', verify = True, cert = None

    def cert_verify(self, conn, url, verify, cert):
        """Verify a SSL certificate. This method should not be called from user
        code, and is only exposed for use when subclassing the
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.
    
        :param conn: The urllib3 connection object associated with the cert.
        :param url: The requested URL.
        :param verify: Either a boolean, in which case it controls whether we verify
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use
        :param cert: The SSL certificate to verify.
        """
        if url.lower().startswith('https') and verify:
    
            cert_loc = None
    
            # Allow self-specified cert location.
            if verify is not True:
                cert_loc = verify
    
            if not cert_loc:
                cert_loc = extract_zipped_paths(DEFAULT_CA_BUNDLE_PATH)
    
            if not cert_loc or not os.path.exists(cert_loc):
>               raise IOError("Could not find a suitable TLS CA certificate bundle, "
                              "invalid path: {}".format(cert_loc))
E               OSError: Could not find a suitable TLS CA certificate bundle, invalid path: /Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/certifi/cacert.pem

/Users/tammylau/conda/envs/dnadoh_dev/lib/python3.8/site-packages/requests/adapters.py:228: OSError
```

Greg to follow-up